### ISSUE TITLE
Enhancement: Allowing plugins to propose models that would be incorporated automatically in the built-in models library

### ISSUE THREAD

**Issue by pcaillou** [gama-platform/gama#670]
```
This enhancement is technically simple to do but could be very interesting : allowing
models defined in plugins (not in the msi.gama.models plugin) to be detected and added
automatically to the built-in library of models. 

To do it properly, we could for example add a new extension to the plugins (like the
gaml extension) that would define a title for the project/folder, and point to the
folder that will become available in GAMA as a project (or a set of projects). 

This request is of course strongly linked with Issue #657 and also Issue #656, and we
need to think globally how to address them. 

```

Original issue reported on code.google.com by `alexis.drogoul` on 2013-10-02 18:05:22

**Comment by AlexisDrogoul** [comment]
I have begun working on this enhancement. Basically, any plugin that declares a `gaml` extension can also have a `models` sub package, at the same level of `gaml`. 

All the models put in this folder (as long as they are GAMA projects, etc.) are contributed to the Models Library, under `Features > 'human readable name of the plugin'`.

This contribution is dynamic: adding a new plugin with this extension will add the new models (maybe a refresh will be necessary, but it can be automated as well) and removing the plugin will remove them. 

The big interest is that plugins can document themselves this way without having to hack the models library.  One more interest of this enhancement will be to be able to distribute models in 'empty' plugins (using the update mechanism). 

Please comment on this if you have any other special requirement (as I'm currently coding it…).
(I add @benoitgaudou @gnoubi @agrignard to the discussion)

**Comment by RCura** [comment]
Having a simple way to "package" models (together with their ressources etc...) as plugins (so far as a plugin isn't too hard to create/maintain/update) would definitely be very usefull for collaborative modelling and sharing models :+1:

**Comment by AlexisDrogoul** [comment]
Well. I am able to import projects from plugins and to add them to a new category ("Plugin Models"). The only requirement for plugins is that (1) they define the extension 'gaml.grammar.addition' in their plugin.xml; (2) they contain a folder called 'models' in which the GAMA projects are located (in sub-folders). Other than that, they do not need any additional files or information.

So, right now, it is possible to define plugins that only contains models and which can be added/exchanged through the regular update mechanisms. 

I'm however still refining what I have coded, to clarify things a bit further (using the "gaml.grammar.addition" extension is a bit odd — it would be better to declare a dedicated extension, e.g. "gaml.models" that would point to a folder containing the projects, which could then be the root of the plugin or any other location inside it, and which could also allow to define a short documentation).

**Comment by AlexisDrogoul** [comment]
I've basically finished to code this enhancement. I have two more things to do:

1) That models can be correctly read from .jar files, which is not completely evident to do (i.e. resources mapped from .jar files are not easy to deal with — so the solution is, either, that the corresponding plugin is declared as not archivable by its feature (like msi.gama.models) or that we decompress at runtime the corresponding contents in a temp directory, but I have yet to implement this).

2) To provide a new extension point (`gama.models`) to allow developers to specify a directory in which to find models. If it is not defined, and if `gaml.grammar.additions` is extended, the BundleLoader will nevertheless look for a folder called `models` in the plugin. 

I've tested the solution by moving some projects in Features to their corresponding plugins. You can see the result in the workspace:

<img width="461" alt="capture 2015-12-30 a 14 40 17" src="https://cloud.githubusercontent.com/assets/579256/12047860/2d2f3f78-af05-11e5-9f36-b4b3565feb91.PNG">
 
Feel free to comment on this change.

**Comment by benoitgaudou** [comment]
Given RCura's remark, I am wondering whether it could be possible, directly in GAMA to right-click on a model project and chose to package and thus create a jar that could be shared.

**Comment by RCura** [comment]
@benoitgaudou : Indeed, would be a very useful feature :) The current mechanism of model import in Gama is quite confusing and not very straight-forward when working with different people.

**Comment by AlexisDrogoul** [comment]
It's a different issue we talk about here (about sharing models) and I propose that you open a new issue with (1) what is confusing/complicated in the current way of sharing models (or maybe what is not documented enough as there are many ways to share models); (2) what (and maybe how) it should be improved.

**Comment by AlexisDrogoul** [comment]
The addition of the new extension point `gama.models` is done and allows to specify a different path for the folder containing model projects. I have also taken this opportunity to deprecate `gaml.grammar.addition` in favor of `gaml.extension` (removing the need to specify a path, since this path is fixed and decided in the annotation processor).

The last thing to check, now, is to verify that GAMA (and all the places that use files : editor, models, viewers, etc.) can correctly access files within .jar archives. I'm actually sure of the contrary, but I need to test where it fails — and, if needed, to implement a workaround (decompression in a temp dir, or necessity to distribute the plugin as non archived).

**Comment by AlexisDrogoul** [comment]
OK. I will commit the changes made so far. It seems that the method used to retrieve resources/files into jar plugins automatically decompresses them so that they are ready to be used in other parts. This needs to be more thoroughly tested, however.

Regarding the parallel discussion on "packaging", these changes allow to distribute and update models (through the update mechanism) in a simple way. Either they are part of GAML projects within GAMA plugins, or they constitute themselves a project/plugin (with minimal information: only the dependency to msi.gama.core + the gama.models extension needs to be provided. See for example the layout of a test plugin I have used. Only circle.gaml is relevant here, the rest is simply packaging.

<img width="337" alt="capture 2016-01-02 a 18 36 58" src="https://cloud.githubusercontent.com/assets/579256/12074782/eb6c37f2-b199-11e5-8eef-b5b5ad0f8518.PNG">

Basically, such a model can perfectly be part of a feature and posted on an update site, so that it can be downloaded in GAMA, and then updated regularly. The thing is that I dont exactly see which benefit it provides compared to an import (as a link) of a folder shared by people on GoogleDrive or DropBox. 

So, for now on, let's consider this feature as a way to bundle sample models with plugins, which was its original intention. I think it will considerably ease the maintenance of the library of models, esp. the ones dedicated to specific features found in plugins.
